Firestore/core/src/firebase/firestore/model/field_mask.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_MASK_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_MASK_H_
+
+#include <initializer_list>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/**
+ * Provides a set of fields that can be used to partially patch a document.
+ * FieldMask is used in conjunction with FieldValue of Object type.
+ *
+ * Examples:
+ *   foo - Overwrites foo entirely with the provided value. If foo is not
+ *       present in the companion FieldValue, the field is deleted.
+ *   foo.bar - Overwrites only the field bar of the object foo. If foo is not an
+ *       object, foo is replaced with an object containing bar.
+ */
+class FieldMask {
+ public:
+  explicit FieldMask(std::initializer_list<FieldPath> list) : fields_{list} {
+  }
+  explicit FieldMask(const std::vector<FieldPath>& fields) : fields_{fields} {
+  }
+  explicit FieldMask(std::vector<FieldPath>&& fields) : fields_{fields} {
+  }
+
+  const std::vector<FieldPath>& fields() const {
```

Comment:
Entirely optional suggestion: instead of exposing the vector, you could add public member functions `begin()` and `end()`. In that case, range based for could work on `FieldMask` itself: `for (const auto& field : mask) {}`. `operator==` could delegate to `std::equal` or `absl::equal`.

Reply:
I'm a fan of exposing `begin() const` and `end() const`.

Delegating to `std::equal` is much more verbose than the more obvious way of delegating to `std::vector::operator==`.

Reply:
done